### ISSUE TITLE
Custom HystrixConcurrencyStrategy doesn't work when using hystrix plugin.

### ISSUE THREAD

**Issue by ScienJus** [apache/skywalking#1892]
Please answer these questions before submitting your issue.

- Why do you submit this issue?
- [ ] Question or discussion
- [x] Bug
- [ ] Requirement
- [ ] Feature or performance improvement

___
### Bug
- Which version of SkyWalking, OS and JRE?

SkyWalking 6.0.0-alpha, hystrix-1.x-plugin 6.0.0-alpha

- What happen?

In spring cloud, people like to programmatically set HystrixConcurrencyStrategy through a delegate pattern, let me just give two examples:

[SecurityContextConcurrencyStrategy in spring-cloud-netflix](https://github.com/spring-cloud/spring-cloud-netflix/blob/6859b7a83588f33c8a946de40cfbab0917de3d05/spring-cloud-netflix-hystrix/src/main/java/org/springframework/cloud/netflix/hystrix/security/HystrixSecurityAutoConfiguration.java#L48-L69)

[SleuthHystrixConcurrencyStrategy in spring-cloud-sleuth](https://github.com/spring-cloud/spring-cloud-sleuth/blob/99d38afc6993685653c102216f964fa1dec8c95d/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/hystrix/SleuthHystrixConcurrencyStrategy.java#L62-L92)

But hystrix plugin will make them unable to work because the result of `HystrixPlugins#getConcurrencyStrategy` is cached in the HystrixConcurrencyStrategyInterceptor, although after that other HystrixConcurrencyStrategy will reset and register itself. Then HystrixConcurrencyStrategyInterceptor will return the wrong result.

We have several ways to fix this problem:

1. Believe that all HystrixConcurrencyStrategy adheres to the delegate pattern, only need to wrap and register SWHystrixConcurrencyStrategyWrapper in the first call.

```
@Override
public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments, Class<?>[] argumentsTypes,
                          Object ret) throws Throwable {
    SWHystrixPluginsWrapperCache wrapperCache = (SWHystrixPluginsWrapperCache) objInst.getSkyWalkingDynamicField();
    if (wrapperCache == null || wrapperCache.getSwHystrixConcurrencyStrategyWrapper() == null) {
        synchronized (objInst) {
            if (wrapperCache == null) {
                wrapperCache = new SWHystrixPluginsWrapperCache();
                objInst.setSkyWalkingDynamicField(wrapperCache);
            }
            if (wrapperCache.getSwHystrixConcurrencyStrategyWrapper() == null) {
                // Return and register wrapper only for the first time
                // try to believe that all other strategies will use the their delegates
                SWHystrixConcurrencyStrategyWrapper strategy = new SWHystrixConcurrencyStrategyWrapper((HystrixConcurrencyStrategy) ret);
                wrapperCache.setSwHystrixConcurrencyStrategyWrapper(strategy);

                registerSWHystrixConcurrencyStrategyWrapper(strategy);

                return strategy;
            }
        }
    }
    return ret;
}

private void registerSWHystrixConcurrencyStrategyWrapper(SWHystrixConcurrencyStrategyWrapper swHystrixConcurrencyStrategyWrapper) {
    // Copy from Spring Cloud Sleuth
    HystrixCommandExecutionHook commandExecutionHook = HystrixPlugins
            .getInstance().getCommandExecutionHook();
    HystrixEventNotifier eventNotifier = HystrixPlugins.getInstance()
            .getEventNotifier();
    HystrixMetricsPublisher metricsPublisher = HystrixPlugins.getInstance()
            .getMetricsPublisher();
    HystrixPropertiesStrategy propertiesStrategy = HystrixPlugins.getInstance()
            .getPropertiesStrategy();
    HystrixPlugins.reset();
    HystrixPlugins.getInstance().registerConcurrencyStrategy(swHystrixConcurrencyStrategyWrapper);
    HystrixPlugins.getInstance()
            .registerCommandExecutionHook(commandExecutionHook);
    HystrixPlugins.getInstance().registerEventNotifier(eventNotifier);
    HystrixPlugins.getInstance().registerMetricsPublisher(metricsPublisher);
    HystrixPlugins.getInstance().registerPropertiesStrategy(propertiesStrategy);
}
```

2. If we don't trust the delegate pattern and must make
the SWHystrixConcurrencyStrategy exists in the delegate chain, it will be difficult, not only to intercept get/reset/register, but also there is no good way to confirm if a class is already on a delegate chain.

What do you think? Do you need a demo project to reproduce this issue?

**Comment by wu-sheng** [comment]
I am not familiar with Hystrix, could you explain more about the option (1). What is `registerSWHystrixConcurrencyStrategyWrapper` do? Look like reset the old way?

**Comment by wu-sheng** [comment]
@ascrutae Could you take part in this? too.

**Comment by ScienJus** [comment]
@wu-sheng Because `registerConcurrencyStrategy` does not allow registration when there is already an implementation, so it must be reset before registration.

```
public void registerConcurrencyStrategy(HystrixConcurrencyStrategy impl) {
    if (!this.concurrencyStrategy.compareAndSet((Object)null, impl)) {
        throw new IllegalStateException("Another strategy was already registered.");
    }
}
```

**Comment by ScienJus** [comment]
A more detailed example:

## Current

1. A custom HystrixConcurrencyStrategy (named Custom) wants to wrap the current HystrixConcurrencyStrategy, so it calls `HystrixPlugin#getConcurrencyStrategy`
2. `HystrixPlugin#getConcurrencyStrategy` loads the default implementation (called Default)
3. HystrixConcurrencyStrategyInterceptor intercepts this method, wraps and caches the result (called SW), at this point:
    - Field `HystrixPlugin.concurrencyStrategy`: Default
    - Intercepted result: SW[Default]
4. Custom wraps SW[Default] with delegate pattern and registers itself with `HystrixPlugin#reset` and `HystrixPlugin#registerConcurrencyStrategy`, at this point:
    - Field `HystrixPlugin.concurrencyStrategy`: Custom[SW[Default]]
5. A new HystrixCommand is created, and get the current implementation by `HystrixPlugin#getConcurrencyStrategy`. Also HystrixConcurrencyStrategyInterceptor intercepts this method, and returns the cached result, at this point:
    - Field `HystrixPlugin.concurrencyStrategy`: Custom[SW[Default]] (correct)
    - Intercepted result: SW[Default] (wrong, Custom is lost)

## Option 1

1. A custom HystrixConcurrencyStrategy (named Custom) wants to wrap the current HystrixConcurrencyStrategy, so it calls `HystrixPlugin#getConcurrencyStrategy`
2. `HystrixPlugin#getConcurrencyStrategy` loads the default implementation (called Default)
3. HystrixConcurrencyStrategyInterceptor intercepts this method, wraps and register itself(called SW), at this point:
    - Field `HystrixPlugin.concurrencyStrategy`: SW[Default]
    - Intercepted result: SW[Default]
4. Custom wraps SW[Default] with delegate pattern and registers itself with `HystrixPlugin#reset` and `HystrixPlugin#registerConcurrencyStrategy`, at this point:
    - Field `HystrixPlugin.concurrencyStrategy`: Custom[SW[Default]]
5. A new HystrixCommand is created, and get the current implementation by `HystrixPlugin#getConcurrencyStrategy`. Also HystrixConcurrencyStrategyInterceptor intercepts this method, but do nothing, at this point:
    - Field `HystrixPlugin.concurrencyStrategy`: Custom[SW[Default]] (correct)
    - Intercepted result: Custom[SW[Default]] (correct)

**Comment by wu-sheng** [comment]
I got your most point. So in this scenario, must use reset then `register*` as the failsafe mechanism. I think this is the right suggestion. How about set a pull request to this, and verify through existing test cases?

I think you have tested the new one in your env, right?

**Comment by ScienJus** [comment]
> I think you have tested the new one in your env, right?

Yes, the hystrix-plugin of master branch is failed in my integration tests, and after adding this patch, it passed.

So I should do:

1. Add an integration test to make the current version fail(I have a simple project to reproduce this problem)
2. Create a pull request, we can review whether it will bring new problems
3. Make sure it passes the existing integration tests

Maybe 1. and 3. can be done in @SkywalkingTest?

**Comment by wu-sheng** [comment]
Yes, all cases are here, https://github.com/SkywalkingTest/skywalking-agent-testcases

**Comment by ScienJus** [comment]
Some other questions:

1. Although I have not encountered it, it seems that HystrixPluginsInterceptor will also have the same issue?
2. If the `SWHystrixConcurrencyStrategy` only needs to be initialized once, do we still need to use ASM?

Looking forward to @ascrutae's advice.

**Comment by wu-sheng** [comment]
> Although I have not encountered it, it seems that HystrixPluginsInterceptor will also have the same issue?

Look like it should, do you have case to try?

> If the SWHystrixConcurrencyStrategy only needs to be initialized once, do we still need to use ASM?

Do you mean SWHystrixConcurrencyStrategyWrapper? I don't think it uses ASM in any way.

**Comment by ScienJus** [comment]
> Look like it should, do you have case to try?

Sure, it's easy to create a case.

> Do you mean SWHystrixConcurrencyStrategyWrapper? I don't think it uses ASM in any way.

In the first option, `HystrixConcurrencyStrategyInterceptor` will only register `SWHystrixConcurrencyStrategyWrapper` once, and never modify the result (the code snippet in the first option), so is the `HystrixConcurrencyStrategyInterceptor`(uses ASM) necessary? Or just need a static initialization method?

**Comment by wu-sheng** [comment]
> In the first option, HystrixConcurrencyStrategyInterceptor will only register SWHystrixConcurrencyStrategyWrapper once, and never modify the result (the code snippet in the first option), so is the HystrixConcurrencyStrategyInterceptor(uses ASM) necessary? Or just need a static initialization method?

If not an interceptor, how to make wrapper work? I am a little confused. In real world, you even can't know, Hystrix exists or not.

**Comment by wu-sheng** [comment]
Any static code blocks should consider about classloader and ClassNotFound issues. That is the key interceptor being guaranteed by agent core.

**Comment by ScienJus** [comment]
> Any static code blocks should consider about classloader and ClassNotFound issues. That is the key interceptor being guaranteed by agent core.

You are right, this is the most important thing.

**Comment by ascrutae** [comment]
After I read this plugin. I found that the root cause of this issue is that this plugin does not take into account the situation of call `reset` method. I think that we just intercept `reset` method and reset the `SWHystrixPluginsWrapperCache` object will be ok. Here is the pseudo code of the new interceptor:
```java
((EnhancedInstance)(HystrixPlugins.getInstance())).setSkyWalkingDynamicField(null);
```

**Comment by ScienJus** [comment]
@ascrutae I also thought about this, but there is also a problem:

If Custom[SW[Default]] reset and register itself, then we clear the cache, next time the interceptor will wrap it again, then it will become SW[Custom[SW[Default]]], does it causes repeated spans?

**Comment by wu-sheng** [comment]
Ha, that really tricky scenario. But make sense.

**Comment by ascrutae** [comment]
Why would there be  `Custom[SW[Default]]`  situation? What I understand is that Hystrix only set the one `HystrixConcurrencyStrategy`, that is mean the `registerConcurrencyStrategy` method executes before the `getConcurrencyStrategy` method or it will use the [default strategy](https://github.com/Netflix/Hystrix/blob/master/hystrix-core/src/main/java/com/netflix/hystrix/strategy/HystrixPlugins.java#L165)，so  `HystrixPluginsInterceptor` wrap the return value of `getConcurrencyStrategy` method and it will get the right way.

The Hystrix get `HystrixConcurrencyStrategy` by `getConcurrencyStrategy` method and clear `HystrixConcurrencyStrategy` by `reset` method, so Hystrix clear, we clear and Hystrix get, we put or get.

**Comment by wu-sheng** [comment]
i think @ScienJus means, if get enhanced strategy and set it back

**Comment by ScienJus** [comment]
@ascrutae @wu-sheng yes, for example in [SleuthHystrixConcurrencyStrategy](https://github.com/spring-cloud/spring-cloud-sleuth/blob/99d38afc6993685653c102216f964fa1dec8c95d/spring-cloud-sleuth-core/src/main/java/org/springframework/cloud/sleuth/instrument/hystrix/SleuthHystrixConcurrencyStrategy.java#L66) , `this.delegate` will be `SW[Default]`, so it will register `Sleuth[SW[Default]]` as itself.

**Comment by wu-sheng** [comment]
I have discussed with Xin. My question is, could your solution fix this? And how.

If not, I and Xin have a proposal, which is dofields scan for every strategy object. If skywalking wrapper found there, assume no need to do more. What do you think?

This solution is not perfect, but should good enough.

**Comment by ScienJus** [comment]
> My question is, could your solution fix this? And how.

My solution based on every HystrixConcurrencyStrategy implements a delegate pattern. If one of them does not, the SWHystrixConcurrencyStrategyWrapper will be lost. Just like if someone doesn't follow this plan, spring cloud slueth will also fail.

> which is dofields scan for every strategy object. If skywalking wrapper found there, assume no need to do more. What do you think?

Like the second option I said:

> If we don't trust the delegate pattern and must make
the SWHystrixConcurrencyStrategy exists in the delegate chain, it will be difficult, not only to intercept get/reset/register, but also there is no good way to confirm if a class is already on a delegate chain.

It will increase the complexity and still can not cover all situations (if the delegate is a List or another wrapper class does not explicitly extend HystrixConcurrencyStrategy ?).

I prefer to follow the existing conventions (how spring cloud does it), but the decision is up to you.

**Comment by wu-sheng** [comment]
To be specific, my question is about the scenario you bring up, if the get a strategy already wrappered, wrap it, then put back. You will still get double wrappers, right?

**Comment by wu-sheng** [comment]
Just like you show up in Spring example. Why your option 1 could avoid?

**Comment by ScienJus** [comment]
in #1901. Just because it is only executed for the first time.

**Comment by wu-sheng** [comment]
I think I finally got your whole point. I am ok with your pr, now.

**Comment by wu-sheng** [comment]
Please submit new test cases, as we discussed.

**Comment by ScienJus** [comment]
@wu-sheng @acurtain 

Today we found another issue about Hystrix in Spring Cloud 2, it can also be shared with you for reference.

https://github.com/spring-cloud/spring-cloud-netflix/issues/3256

Spring bean injection breaks the delegate chain and causes other HystrixConcurrencyStrategy(like SWHystrixConcurrencyStrategy) to be lost.

**Comment by ScienJus** [comment]
@wu-sheng We don't need to track it because this is a spring cloud bug, I just show you some bad design :)

**Comment by wu-sheng** [comment]
Oh, interesting. Thank for bringing this to the community.

**Comment by sail-y** [comment]
I got this problem too, I custom my own HystrixConcurrencyStrategy but it doesn't work.I debug the code `HystrixContextScheduler.java` in 62 line,  found that my own HystrixConcurrencyStrategy's instance have always been changed to an intance of SWHystrixConcurrencyStrategyWrapper.

**Comment by wu-sheng** [comment]
@sail-y yes. That change is expected to happens. Other, trace context breaks.

**Comment by sail-y** [comment]
@wu-sheng #1901 included in 6.0.0-alpha already? I tested it yesterday and the bug still showing, my own CustomStrategy is lost.

**Comment by wu-sheng** [comment]
Based on the milestone, no, it should be included in 6.0.0-beta release.

**Comment by wu-sheng** [comment]
#1901 is only included in master codes. If you want, could package the distribution by yourself.
